src/corehost/cli/hostpolicy/hostpolicy.cpp

```diff
@@ -58,30 +66,53 @@ namespace
         return StatusCode::Success;
     }
 
-    int create_hostpolicy_context(
+    int get_or_create_hostpolicy_context(
         hostpolicy_init_t &hostpolicy_init,
         const arguments_t &args,
         bool breadcrumbs_enabled,
-        std::shared_ptr<hostpolicy_context_t> &context)
+        hostpolicy_context_t **context)
     {
+        hostpolicy_context_t *existing_context;
+        {
+            std::unique_lock<std::mutex> lock{ g_context_lock };
+            g_context_cv.wait(lock, [] { return !g_context_initializing.load(); });
+
+            existing_context = g_context.get();
+            if (existing_context == nullptr)
+                g_context_initializing.store(true);
+        }
+
+        if (existing_context != nullptr)
+        {
+            // [TODO] Validate the current context is acceptable for this request
+            trace::info(_X("Host context has already been initialized"));
+            *context = existing_context;
+            return StatusCode::CoreHostAlreadyInitialized;
+        }
 
         std::unique_ptr<hostpolicy_context_t> context_local(new hostpolicy_context_t());
         int rc = context_local->initialize(hostpolicy_init, args, breadcrumbs_enabled);
         if (rc != StatusCode::Success)
-            return rc;
+        {
+            {
+                std::lock_guard<std::mutex> lock{ g_context_lock };
+                g_context_initializing.store(false);
+            }
 
-        context = std::move(context_local);
+            g_context_cv.notify_all();
```

Comment:
I'm trying to wrap my head around the synchronization design here:
* In hostfxr we want to block initialization of secondary context until after the runtime is initialized.
* In hostpolicy (here) we use similar mechanism, but block only until the hostpolicy is initialzied (computed the properties and so on), but before the runtime is loaded.
* What is the purpose of this synchronization in hostpolicy?
  * In theory we don't need it since the one in hostfxr should cover the hostpolic as well.
  * I can understand trying to be careful and make the API in hostpolicy itself robust (after all it is a public API in a way).
  * But in that case I would assume similar scopes of synchronization between hostfxr and hostpolicy... I guess I'm missing something.

Reply:
I put the synchronization in hostpolicy as well for the reason you state of it being an exposed public API itself. It should be blocking until the runtime is loaded or hostpolicy initialization fails (the case in the above code). I think I'm missing how it is getting unblocked after hostpolicy initialized but before the runtime is loaded?

Reply:
You're right - I completely ignored the fact that this one is only for the error cases. Makes sense now - thanks.